protocol/contracts/farm/facets/MarketplaceFacet/MarketplaceFacet.sol

```diff
@@ -0,0 +1,292 @@
+
+        require(bean().balanceOf(msg.sender) >= amountBeans, "Marketplace: Not enough beans to submit buy offer.");
+        require(pricePerPod > 0, "Marketplace: Price must be greater than 0");
+
+        uint256 boughtBeanAmount = LibMarket.buyExactTokens(buyBeanAmount);
+
+        uint256 amountBeansUsing = boughtBeanAmount + amountBeans;
+        require(amountBeansUsing > 0, "Marketplace: Must offer to buy non-zero amount");
+
+        uint256 amount = (amountBeansUsing  * 1000000) / pricePerPod;
+
+        s.buyOffers[s.buyOfferIndex].amount = amount;
+        s.buyOffers[s.buyOfferIndex].price = pricePerPod;
+        s.buyOffers[s.buyOfferIndex].maxPlaceInLine = maxPlaceInLine;
+        s.buyOffers[s.buyOfferIndex].owner = msg.sender;
+
+        bean().transferFrom(msg.sender, address(this), amountBeans);
+
+        emit BuyOfferCreated(s.buyOfferIndex, msg.sender, amount, pricePerPod, maxPlaceInLine);
+
+        console.log('bean().balanceOf(msg.sender) after',bean().balanceOf(address(this)));
+
+    }
+
+
+    function sellToBuyOffer(uint256 plotIndex, uint24 buyOfferIndex, uint232 amount) public  {
+        
+        require(s.a[msg.sender].field.plots[plotIndex] >= 0, "Marketplace: Plot  not owned by user.");
+
+
+
+        uint232 harvestable = uint232(s.f.harvestable);
+        require(plotIndex >= harvestable, "Marketplace: Cannot send harvestable plot");
+
+        Storage.BuyOffer storage buyOffer = s.buyOffers[buyOfferIndex];
+        uint256 placeInLine = plotIndex + amount - harvestable;
+
+        require(placeInLine <= buyOffer.maxPlaceInLine, "Marketplace: Plot too far in line");
+        require(buyOffer.price > 0, "Marketplace: Buy Offer does not exist");
+        require(amount <= buyOffer.amount, "Marketplace: Buy Offer has insufficient amount.");
+
+        uint256 costInBeans = (buyOffer.price * amount) / 1000000;
+
+        bean().transfer(msg.sender, costInBeans);
+
+        // If plot being transferred was previously listed,
+        // Update Index / Delete Listing Accordingly 
+        if (s.listedPlots[plotIndex].price > 0){
+            _fillListing(plotIndex, amount, buyOffer.owner,msg.sender, true);
+
+            // Storage.Listing storage listing = s.listedPlots[plotIndex];
+
+            // uint256 listingAmount = listing.amount;
+            // if (listingAmount == 0){
+            //     listingAmount = s.a[msg.sender].field.plots[plotIndex];
+            // }
+            // if (amount == listingAmount){
+            //     delete s.listedPlots[plotIndex];
+            // }
+            // else{
+            //     s.listedPlots[plotIndex.add(amount)] = listing;
+            //     s.listedPlots[plotIndex.add(amount)].amount = listingAmount - amount;
+            //     delete s.listedPlots[plotIndex];
+            // }
+            // emit ListingFilled(buyOffer.owner, msg.sender, plotIndex, listing.price, amount);
+
+
+        }
+
+        buyOffer.amount = buyOffer.amount - amount;
```

Comment:
So we should discuss what the best thing to do here -> this line + line 186 is the same thing as using SafeMath. Throughout the project we have used SafeMath, but this is fine too.